app/util/cron.py

```diff
@@ -0,0 +1,69 @@
+import atexit
+import datetime
+import json
+import time
+
+from app import app
+from sqlalchemy import func, exc
+from app.model.email import Email, EmailStatus
+from app.util.db import db
+from app.util.mailer import mailer
+from apscheduler.schedulers.background import BackgroundScheduler
+from apscheduler.triggers.interval import IntervalTrigger
+
+JOB_INTERVAL_SEC = 10
+RETRY_INTERVAL_MIN = 10
+
+
+def send_mail():
+    with app.app_context():
+        mails_new = Email.query \
+            .filter(
+            Email.status == False,
+            Email.counter == 0
+        ).all()
+        mails_failed = Email.query \
+            .filter(
+            Email.status == False,
+            Email.last_sent_at < (datetime.datetime.utcnow() - datetime.timedelta(minutes=RETRY_INTERVAL_MIN)),
+            Email.counter < 4
+        ).all()
+
+        for mail_message in mails_new + mails_failed:
+            app.logger.info("Sending email #" + str(mail_message.id))
+            recipients = [mail_message.mail_to]
+            response = {}
+            if mail_message.mail_cc is not None and mail_message.mail_cc:
+                recipients += [mail_message.mail_cc]
+            if mail_message.mail_bcc is not None and mail_message.mail_bcc:
+                recipients += [mail_message.mail_bcc]
+
+            try:
+                response = mailer.send_message(mail_message.mail_from, recipients, mail_message.body)
+            except Exception as e:
+                app.logger.error(str(e))
```

Comment:
An exception is swallowed and the function proceeds as if nothing happened. The logic below does not seem to expect that.

Reply:
fixed, i guess three tries for mail sending maybe not enough in case of exception.